src/files-breadcrumbs.tsx

```diff
@@ -51,6 +60,143 @@ function useHostname() {
     return hostname;
 }
 
+function BookmarkButton({ path }: { path: string[] }) {
+    const [isOpen, setIsOpen] = React.useState(false);
+    const [user, setUser] = React.useState<cockpit.UserInfo | null>(null);
+    const [bookmarks, setBookmarks] = React.useState<string[]>([]);
+    const [bookmarkHandle, setBookmarkHandle] = React.useState<cockpit.FileHandle<string> | null>(null);
+
+    const { addAlert, cwdInfo } = useFilesContext();
+
+    const currentPath = path.join("/") || "/";
+    const defaultBookmarks = [];
+    if (user?.home)
+        defaultBookmarks.push({ name: _("Home"), loc: user?.home }); // TODO: add trash
+
+    useInit(async () => {
+        const user_info = await cockpit.user();
+        setUser(user_info);
+
+        const handle = cockpit.file(`${user_info.home}/.config/gtk-3.0/bookmarks`);
+        setBookmarkHandle(handle);
+
+        handle.watch((content) => {
+            if (content !== null) {
+                setBookmarks(content.split("\n").filter(line => line.startsWith("file://"))
+                        .map(line => {
+                            // Drop everything after the space, we don't show renames
+                            line = line.replace(/\s.*/, '');
+
+                            // Drop the file:/// prefix
+                            line = line.replace('file://', '');
+
+                            // Nautilus decodes urls as paths can contain spaces
+                            return line.split('/').map(part => decodeURIComponent(part))
+                                    .join('/');
+                        }));
+            } else {
+                setBookmarks([]);
+            }
+        });
+
+        return [handle];
+    });
+
+    const saveBookmark = async () => {
+        cockpit.assert(user !== null, "user is null while saving bookmarks");
+        cockpit.assert(bookmarkHandle !== null, "bookmarkHandle is null while saving bookmarks");
+        const bookmark_file = basename(bookmarkHandle.path);
+        const config_dir = bookmarkHandle.path.replace(bookmark_file, "");
+
+        try {
+            await cockpit.spawn(["mkdir", "-p", config_dir]);
+        } catch (err) {
+            const exc = err as cockpit.BasicError; // HACK: You can't easily type an error in typescript
+            addAlert(_("Unable to create bookmark directory"), AlertVariant.danger, "bookmark-error",
+                     exc.message);
+            return;
+        }
+
+        try {
+            await bookmarkHandle.modify((old_content: string) => {
+                const newBoomark = path.map(part => encodeURIComponent(part))
+                        .join('/');
+                if (bookmarks.includes(currentPath)) {
+                    return old_content.split('\n').filter(line => !line.startsWith(`file://${newBoomark}`))
```

Comment:
This will fail to detect an entry in the bookmarks file which had a different encoding than the one we'd have selected.  We'll match our `bookmarks.includes()` check above (because we parsed the entries out for that one) but fail to match on our encoding.

I think it might be interesting to reuse the parser code from our `.watch()` handler, which is quite good.  It could find the line that matches, and then we could remove that one from the list.  We know we're sure to find it, because it was the line that resulted in us having the bookmark in the first place.

That could work as a filter as well, something like

```ts
  return old_content.split('\n').filter(line => parse_uri(line) !== currentPath).join('\n')
```

This is probably worth adding a test case over.

Reply:
Also: this code is slightly weird.  It considers the zero bytes after the last newline in the file as if they were a bookmark entry.  In that case, it's zero bytes, so it doesn't match, but ... it feels weird.  It might be nice to have some "proper" line-splitting code here.

I wonder if you'd consider using a simple "syntax" object on your file handle to take care of the line splitting and re-joining.  Then you could just interact directly with lines in both the watcher and the modifier.

Reply:
> I think it might be interesting to reuse the parser code from our .watch() handler, which is quite good. It could find the line that matches, and then we could remove that one from the list. We know we're sure to find it, because it was the line that resulted in us having the bookmark in the first place.

Took me a while to parse this, the `parse_uri` approach seems to be fine. I'm happy to make those changes.

> Also: this code is slightly weird. It considers the zero bytes after the last newline in the file as if they were a bookmark entry. In that case, it's zero bytes, so it doesn't match, but ... it feels weird. It might be nice to have some "proper" line-splitting code here.

The zero byte case, I mean, in reality it is hardly a problem? I have no idea what you mean with "proper" line splitting, just trimming out the junk?

> I wonder if you'd consider using a simple "syntax" object on your file handle to take care of the line splitting and re-joining. Then you could just interact directly with lines in both the watcher and the modifier.

Like how would that help or make things easier? Its still going to be JavaScript which does the parsing and it would mean we have to re-construct the whole file when modifying from the state we have, which is imo not what we want.

We don't even care about a zero byte in /etc/passwd, why is this file so special? What we do do is trim whitespace (newline or [whitespace character](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#white_space)). Which we can also implement.

```
    content = (content || "").trim();
```

Reply:
> proper line splitting

Like something that converts `"a\nb\n"` to `['a', 'b']` rather than `['a', 'b', '']` and then `['a', 'b']` to `"a\nb\n" rather than `"a\nb"`.